design/src/rfcs/rfc0012_constraint_traits.md

```diff
@@ -0,0 +1,424 @@
+  Smithy specification),
+* the client may have been generated with an older version of the model; and
+* the most recent model version might have lifted some constraints.
+
+On the other hand, server SDKs constitute _the_ source of truth for the
+service's behavior, so they interpret the model in all its strictness.
+
+The Smithy spec defines 8 constraint traits:
+
+* [`enum`],
+* [`idRef`],
+* [`length`],
+* [`pattern`],
+* [`private`],
+* [`range`],
+* [`required`]; and
+* [`uniqueItems`].
+
+The `idRef` and `private` traits are enforced at SDK generation time by the
+[`awslabs/smithy`] libraries and bear no relation to generated Rust code.
+
+The only constraint trait enforcement that is generated by smithy-rs clients
+should be and is the `enum` trait, which renders Rust `enum`s.
+
+The `required` trait is already and only enforced by smithy-rs servers since
+[#1148](https://github.com/awslabs/smithy-rs/pull/1148).
+
+That leaves 4 traits: `length`, `pattern`, `range`, and `uniqueItems`.
+
+[Constraint traits]: https://awslabs.github.io/smithy/1.0/spec/core/constraint-traits.html
+[awslabs/smithy#1039]: https://github.com/awslabs/smithy/issues/1039
+[`enum`]: https://awslabs.github.io/smithy/1.0/spec/core/constraint-traits.html#enum-trait
+[`idRef`]: https://awslabs.github.io/smithy/1.0/spec/core/constraint-traits.html#idref-trait
+[`length`]: https://awslabs.github.io/smithy/1.0/spec/core/constraint-traits.html#length-trait
+[`pattern`]: https://awslabs.github.io/smithy/1.0/spec/core/constraint-traits.html#pattern-trait
+[`private`]: https://awslabs.github.io/smithy/1.0/spec/core/constraint-traits.html#private-trait
+[`range`]: https://awslabs.github.io/smithy/1.0/spec/core/constraint-traits.html#range-trait
+[`required`]: https://awslabs.github.io/smithy/1.0/spec/core/constraint-traits.html#required-trait
+[`uniqueItems`]: https://awslabs.github.io/smithy/1.0/spec/core/constraint-traits.html#uniqueItems-trait
+[`awslabs/smithy`]: https://github.com/awslabs/smithy
+
+Implementation
+--------------
+
+This section addresses how to implement and enforce the `length`, `pattern`,
+`range`, and `uniqueItems` traits. We will use the `length` trait applied to a
+`string` shape as a running example. The implementation of this trait mostly
+carries over to the other three.
+
+### Example implementation for the `length` trait
+
+Consider the following Smithy model:
+
+```smithy
+@length(min: 1, max: 69)
+string NiceString
+```
+
+The central idea to the implementation of constraint traits is: [parse, don't
+validate]. Instead of code-generating a Rust `String` to represent `NiceString`
+values and perform the _validation_ at request deserialization, we can
+[leverage Rust's type system to guarantee domain invariants]. We can generate a
+wrapper [tuple struct] that _parses_ the string's value and is "tight" in the
+set of values it can accept:
+
+```rust
+pub struct NiceString(String);
```

Comment:
it's worth calling out that adding a constraint trait is now a breaking change for servers since it adds this type wrapper. I haven't thought about the trade off deeply, but it seems worth mentioning / considering since it will make it a bit more of a hassle to add constraints retroactively